File: Java/102_Binary_Tree_Level_Order_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList();
        if(root == null ){
            return result;
        }
        Queue<TreeNode> q = new LinkedList();
        q.offer(root);
        while(!q.isEmpty()){
            List<Integer> level = new ArrayList();
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode curr = q.poll();
                level.add(curr.val);
                if(curr.left != null){
                    q.add(curr.left);
                }
                if(curr.right != null){
                    q.add(curr.right);
                }
            }
            result.add(level);
        }
        return result;
    }
}

//DFS version
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> lists = new ArrayList<List<Integer>>();
        helper(lists, root, 0);
        return lists;
    }
    public void helper(List<List<Integer>>lists, TreeNode root , int height){
        if(root == null) return;
        if(height == lists.size()){
            lists.add(new ArrayList<>());
        }
        lists.get(height).add(root.val);
        helper(lists, root.left, height+1);
        helper(lists, root.right, height+1);
    }
}

File: Python/102_binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from ensurepip import version


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        if not root:
            return levels
        
        self.helper(root, 0, levels)
        return levels
        
    def helper(self, node, level, levels):

        if len(levels) == level:
            levels.append([])

        levels[level].append(node.val)

        if node.left:
            self.helper(node.left, level + 1, levels)
        if node.right:
            self.helper(node.right, level + 1, levels)
            
# ==============================
# queue version


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        if not root:
            return levels
        
        level = 0
        queue = deque([root,])
        while queue:
            levels.append([])
            levels_len = len(queue)
            for i in range(levels_len):
                node = queue.popleft()
                levels[level].append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)         
            level +=1
        
        return levels
        